### PR TITLE
[Workflow] Add payment state machine config

### PR DESCRIPTION
**Pull request by NoResponseMate**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13 |
| Bug fix?        | no                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no |
| Related tickets | -                      |
| License         | MIT                                                          |

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Tests/Functional/StateMachine/PaymentWorkflowTest.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        $this->payment = new Payment();
+        $this->payment->setOrder($this->order);
+    }
+
+    /**
+     * @test
+     *
+     * @dataProvider availableTransitions
+     */
+    public function it_applies_all_available_transitions_for_create_state(
```
Comment (Wojdylak) [inline src/Sylius/Bundle/CoreBundle/Tests/Functional/StateMachine/PaymentWorkflowTest.php]:
```suggestion
    public function it_applies_all_available_transitions(
```

**Review by Wojdylak — DISMISSED** [review]

File: src/Sylius/Bundle/CoreBundle/EventListener/Workflow/Payment/ProcessOrderPaymentListener.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace Sylius\Bundle\CoreBundle\EventListener\Workflow\Payment;
+
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Order\Processor\OrderProcessorInterface;
+use Symfony\Component\Workflow\Event\CompletedEvent;
+use Webmozart\Assert\Assert;
+
+final class ProcessOrderPaymentListener
```
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/EventListener/Workflow/Payment/ProcessOrderPaymentListener.php]:
```suggestion
final class ProcessOrderListener
```

File: src/Sylius/Bundle/PaymentBundle/Resources/config/app/state_machine.yml [hunk 3]
```diff
@@ -14,7 +14,7 @@ winzou_state_machine:
             completed: ~
             failed: ~
             cancelled: ~
-            void: ~
+            unknown: ~
```
Comment (diimpp) [inline src/Sylius/Bundle/PaymentBundle/Resources/config/app/state_machine.yml]: Should this be mentioned in UPGRADE? Though realistically it won't affect anyone who uses full sylius/sylius and reads UPGRADE.

File: src/Sylius/Bundle/PaymentBundle/Resources/config/app/state_machine.yml [hunk 4]
```diff
@@ -37,4 +37,4 @@ winzou_state_machine:
                 to: refunded
             void:
                 from: [new, processing]
-                to: void
+                to: unknown
```
Comment (diimpp) [inline src/Sylius/Bundle/PaymentBundle/Resources/config/app/state_machine.yml]: On separate note, `void` is valid transition `from [authorized] to voided`, personally I would consider `unknown` a thing to fix instead.

Let's think about it for 2.0.

File: src/Sylius/Bundle/PaymentBundle/Resources/config/workflow/state_machine.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
+framework:
```
Comment (diimpp) [inline src/Sylius/Bundle/PaymentBundle/Resources/config/workflow/state_machine.yaml:1]: Winzou and workflow SMs missing `authorize` transition and `authorized` state. Both of them are defined at interfaces, should they be added to those definition now? Personally I would prefer to have them if they are defined.

File: src/Sylius/Bundle/PaymentBundle/Resources/config/app/state_machine.yml (hunk 3 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/PaymentBundle/Resources/config/app/state_machine.yml]: #15422

File: src/Sylius/Bundle/PaymentBundle/Resources/config/workflow/state_machine.yaml (hunk 5 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/PaymentBundle/Resources/config/workflow/state_machine.yaml:1]: #15423 I added in both place

File: src/Sylius/Bundle/CoreBundle/Resources/config/services/listeners/workflow/payment.xml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ For the full copyright and license information, please view the LICENSE
+ file that was distributed with this source code.
+
+-->
+
+<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+    <services>
+        <defaults public="false" />
+
+        <service id="Sylius\Bundle\CoreBundle\EventListener\Workflow\Payment\ResolveOrderPaymentStateListener">
```
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services/listeners/workflow/payment.xml:24]: Unimportant nitpick, but just noticed, that service definitions are not in alphabetical order

**Review by diimpp — DISMISSED** [review]

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/workflow/sylius_payment.yaml
Comment (jakubtobiasz) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/workflow/sylius_payment.yaml:1]: Why do we need this, as it's almost the same as the configuration from `PaymentBundle` 🤨?

File: src/Sylius/Bundle/CoreBundle/spec/EventListener/Workflow/Payment/ProcessOrderListenerSpec.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    function it_processes_order(
+        OrderProcessorInterface $orderProcessor,
+        PaymentInterface $payment,
+        OrderInterface $order,
+    ): void {
+        $payment->getOrder()->willReturn($order);
+
+        $orderProcessor->process($order)->shouldBeCalled();
+
+        $this->__invoke(new CompletedEvent($payment->getWrappedObject(), new Marking()));
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/EventListener/Workflow/Payment/ProcessOrderListenerSpec.php:57]:
```suggestion
        $this(new CompletedEvent($payment->getWrappedObject(), new Marking()));
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Jan! :tada:

**Pull request merged**